Fix setting attached properties from Styles.

grokys opened the pull request:
This PR fixes the resolution of properties in style setters:

- Updates our parsing of properties to use a custom parser rather than a regex
- Correctly looks up attached properties
- Correctly handles XAML namespaces

I wanted to make trying to set a property that isn't defined on a control throw an exception at XAML parse time, but the animation system relies on being able to assign properties that don't exist to a setter, so had to change that to be a warning.

Fixes #1764

grokys committed:
Added some unit tests for #1764.

grokys committed:
Added PropertyParser.

So we don't need to use a regex to parse property strings.

grokys committed:
Fix a few issues with AvaloniaPropertyTypeConverter

- Don't use regex to parse property strings, use `PropertyParser`
- Handle XAML namespaces on attached properties

Fixes #1764

grokys committed:
Disallow setting unregistered properties.

grokys commented on src/Markup/Avalonia.Markup/Markup/Parsers/Reader.cs:
```diff
@@ -5,7 +5,7 @@
 namespace Avalonia.Markup.Parsers
 {
-    internal class Reader
+    public class Reader
```
@jkoritzinsky I made this class public so it could be used in Avalonia.Markup.Xaml for parsing property strings. Given that, should we rename it? Maybe `CharacterReader` or something? `Reader` feels a bit vague.

grokys replied:
In fact maybe it should also be moved into `Avalonia.Base\Utilities` alongside `StringTokenizer`?

jkoritzinsky replied:
We could rename and move! That works for me.

grokys replied:
Done. I also moved `IdentifierParser`.

danwalmsley reviewed (approved):
Tested and seems to set properties from styles now 👍

Commit:
Merge branch 'master' into fixes/1764-attached-property-setter

grokys commented:
Thanks for the review @danwalmsley - I'd like to get feedback on my question from @jkoritzinsky before we merge though.

jkoritzinsky reviewed (approved):
LGTM other than the rename/move we talked about!

grokys committed:
Don't disallow setting unregistered properties.

The changes made to the animation system in #1768 currently needs to be able to set any property on any object in order for animations on transforms to work.

grokys committed:
Reader -> CharacterReader

And moved `CharacterReader` and `IdentifierParser` into Avalonia.Utilities.
